src/stableswap/math.rs

```diff
@@ -0,0 +1,429 @@
+use crate::types::Balance;
+
+/// Calculating amount to be received from the pool given the amount to be sent to the pool and both reserves.
+/// N - number of iterations to use for Newton's formula
+pub fn calculate_out_given_in<const N: u8>(
+    reserve_in: Balance,
+    reserve_out: Balance,
+    amount_in: Balance,
+    amplification: Balance,
+    precision: Balance,
+) -> Option<Balance> {
+    let ann = two_asset_pool_math::calculate_ann(amplification)?;
+    let new_reserve_out =
+        two_asset_pool_math::calculate_y_given_in::<N>(amount_in, reserve_in, reserve_out, ann, precision)?;
+    reserve_out.checked_sub(new_reserve_out)
+}
+
+/// Calculating amount to be sent to the pool given the amount to be received from the pool and both reserves.
+/// N - number of iterations to use for Newton's formula
+pub fn calculate_in_given_out<const N: u8>(
+    reserve_in: Balance,
+    reserve_out: Balance,
+    amount_out: Balance,
+    amplification: Balance,
+    precision: Balance,
+) -> Option<Balance> {
+    let ann = two_asset_pool_math::calculate_ann(amplification)?;
+    let new_reserve_in =
+        two_asset_pool_math::calculate_y_given_out::<N>(amount_out, reserve_in, reserve_out, ann, precision)?;
+    new_reserve_in.checked_sub(reserve_in)
+}
+
+pub(crate) mod two_asset_pool_math {
+    use super::Balance;
+    use crate::to_u256;
+    use num_traits::Zero;
+    use primitive_types::U256;
+
+    pub(crate) fn calculate_ann(amplification: Balance) -> Option<Balance> {
+        (0..2).try_fold(amplification, |acc, _| acc.checked_mul(2u128))
+    }
+
+    #[inline]
+    fn abs_diff(d0: U256, d1: U256) -> Option<U256> {
+        if d1 > d0 {
+            d1.checked_sub(d0)
+        } else {
+            d0.checked_sub(d1)
+        }
+    }
+
+    /// Calculate `d` so the Stableswap invariant does not change.
+    ///
+    /// Note: this works for two asset pools only!
+    ///
+    /// This is solved using newtons formula by iterating the following equation until convergence.
+    ///
+    /// dn+1 = (ann * S + n * Dp) * dn) / ( (ann -1) * dn + (n+1) * dp)
+    ///
+    /// where
+    ///
+    /// S = sum(xp)
+    /// dp = d^n+1 / prod(sp)
+    ///
+    /// if (dn+1 - dn) <= precision - converged successfully.
+    ///
+    /// Parameters:
+    /// - `xp`: reserves of asset a and b.
+    /// - `ann`: amplification coefficient multiplied by `2^2` ( number of assets in pool)
+    /// - `precision`:  convergence precision
+    pub(crate) fn calculate_d<const N: u8>(xp: &[Balance; 2], ann: Balance, precision: Balance) -> Option<Balance> {
+        let two_u256 = to_u256!(2_u128);
+        let n_coins = two_u256;
```

Comment:
we can use just one variable

Reply:
yes and no. two is actually needed later on. And n_coins is number of coins. 

It is good to see in the actual calculation that it is number of coins rather than just "two".